security/Member.php

```diff
@@ -520,36 +526,56 @@ public static function autoLogin() {
 			return;
 		}
 
-		list($uid, $token) = explode(':', Cookie::get('alc_enc'), 2);
-
-		$member = DataObject::get_by_id("Member", $uid);
-
-		// check if autologin token matches
-		if($member) {
-			$hash = $member->encryptWithUserSettings($token);
-			if(!$member->RememberLoginToken || $member->RememberLoginToken !== $hash) {
-				$member = null;
+		if(strpos(Cookie::get('alc_enc'), ':') && Cookie::get('alc_device') && !Session::get("loggedInAs")) {
+			list($uid, $token) = explode(':', Cookie::get('alc_enc'), 2);
+			$deviceID = Cookie::get('alc_device');
+
+			$member = Member::get()->byId($uid);
+
+			$rememberLoginHash = null;
+
+			// check if autologin token matches
+			if($member) {
+				$hash = $member->encryptWithUserSettings($token);
+				$rememberLoginHash = RememberLoginHash::get()
+					->filter(array(
+						'MemberID' => $member->ID,
+						'DeviceID' => $deviceID,
+						'RememberLoginHash' => $hash
+					))->First();
+				if(!$rememberLoginHash) {
+					$member = null;
+				} else {
+					// Check for expired token
+					$expiryDate = new DateTime($rememberLoginHash->ExpiryDate);
```

Comment:
I think keeping expired tokens can be desirable as it allows retrospective inspection of logged in devices, assurances that tokens WERE expired for a device and so on.